llvm-general/test/LLVM/General/Test/Linking.hs

```diff
@@ -59,9 +59,9 @@ tests = testGroup "Linking" [
         ]      
 
     Module { moduleDefinitions = defs } <- withContext $ \context -> 
-      withModuleFromAST' context ast0 $ \m0 ->
-        withModuleFromAST' context ast1 $ \m1 -> do
-          runExceptT $ linkModules False m0 m1
-          moduleAST m0
+      withModuleFromAST' context ast0 $ \dest -> do
+        src <- failInIO $ astToFFIModule context ast1
+        failInIO $ linkModules dest src
```

Comment:
I’ve gone for a separate type `ModuleRef` rather than `Module` which is basically your ioref suggestion. The reason for this are a) it makes it explicit in the type what can happen to the module b) we don’t introduce unnecessary overhead where we don’t need it (which is everywhere except link) and c) the only way to make this API safe (for users) is to remove the possiblity to do anything with the module after it has been deleted. By not exposing the `getModule` function the user loses that possibility and it should be (reasonably) safe. Given that the only reason why you would want ModuleRef is to be able to call functions that destroy the module this is a reasonable thing to do. 

However it removes the possibilty to first convert the module and then call some other functions before finally destroying it. I am not sure how important this is, but if you think we need this I would just expose `getModule` and put a warning sign on it.